theano/sandbox/cpu_corr/tests/test_conv.py

```diff
@@ -0,0 +1,388 @@
+    mode = None
+    dtype = theano.config.floatX
+
+    def setUp(self):
+        super(TestConv2D, self).setUp()
+        self.input = T.tensor4('input', dtype=self.dtype)
+        self.input.name = 'default_V'
+        self.filters = T.tensor4('filters', dtype=self.dtype)
+        self.filters.name = 'default_filters'
+        if not conv.imported_scipy_signal and theano.config.cxx == "":
+            raise SkipTest("conv2d tests need SciPy or a c++ compiler")
+
+    def validate(self, image_shape, filter_shape,
+                 border_mode='valid', subsample=(1, 1),
+                 N_image_shape=None, N_filter_shape=None,
+                 input=None, filters=None,
+                 verify_grad=True, should_raise=False):
+        """
+        :param image_shape: The constant shape info passed to conv2d.
+        :param filter_shape: The constant shape info passed to conv2d.
+
+        :param N_image_shape: None(default to image_shape) or tuple of
+                              4 elements with the shape of the input image
+
+        :param N_filter_shape: None(default to filter_shape) or tuple
+                               of 4 elements with the shape of the
+                               input filter
+
+        """
+        if N_image_shape is None:
+            N_image_shape = [T.get_scalar_constant_value(T.
+                as_tensor_variable(x)) for x in image_shape]
+        if N_filter_shape is None:
+            N_filter_shape = [T.get_scalar_constant_value(T.
+                as_tensor_variable(x)) for x in filter_shape]
+
+        if input is None:
+            input = self.input
+        if not filters:
+            filters = self.filters
+
+        ############# THEANO IMPLEMENTATION ############
+
+        # we create a symbolic function so that verify_grad can work
+        def sym_CpuCorrMM(input, filters):
+            # define theano graph and function
+            input.name = 'input'
+            filters.name = 'filters'
+            rval =  conv.CpuCorrMM(border_mode, subsample)(input, filters)
+            rval.name = 'conv_output'
+            return rval
+
+        output = sym_CpuCorrMM(input, filters)
+        output.name = 'CpuCorrMM()(%s,%s)' % (input.name, filters.name)
+        theano_conv = theano.function([input, filters], output, mode=self.mode)
+
+        # initialize input and compute result
+        image_data = numpy.random.random(N_image_shape).astype(self.dtype)
+        filter_data = numpy.random.random(N_filter_shape).astype(self.dtype)
+        try:
+            theano_output = theano_conv(image_data, filter_data)
+        except ValueError:
+            if not should_raise:
+                raise
+            return
+        else:
+            if should_raise:
+                raise Exception(
+                "ConvOp should have generated an error")
+
+        ############# REFERENCE IMPLEMENTATION ############
+        # Testing correlation, not convolution. Reverse filters.
```

Comment:
Since this is a correlation and the test is for a convolution, I just flip the filter shapes here.